source/binding/AssignmentExpressions.cpp

```diff
@@ -337,8 +337,14 @@ Expression& Expression::convertAssignment(const BindContext& context, const Type
 Expression& AssignmentExpression::fromSyntax(Compilation& compilation,
                                              const BinaryExpressionSyntax& syntax,
                                              const BindContext& context) {
-    if (!context.flags.has(BindFlags::AssignmentAllowed)) {
-        if (!context.flags.has(BindFlags::NonProcedural) &&
+
+    bool isNonBlocking = syntax.kind == SyntaxKind::NonblockingAssignmentExpression;
+    bool nonBlockInFinal = context.flags.has(BindFlags::Final) && isNonBlocking;
+
+    if (!context.flags.has(BindFlags::AssignmentAllowed) || nonBlockInFinal) {
```

Comment:
I think this would be a little bit more clear if you just check nonBlockInFinal in a separate if condition; there isn't much reason to combine them into one here since you just check the condition again inside.